### PR TITLE
Modifying the TextureList to use EventBrowser-style chooser.

### PR DESCRIPTION
**Pull request by kevin-mccullough**
## Description
This modifies the Texture List to use the same style column chooser as the Event Browser.  It also adds the ability to sort the columns.

### CODE REVIEW

File: qrenderdoc/Code/QRDUtils.cpp
Comment (baldurk) [inline qrenderdoc/Code/QRDUtils.cpp]: Please don't modify the copyright statement in any files. As per the contributing documentation, any contributions must be your own work and submitting a pull request assigns the copyright to them to me.

File: qrenderdoc/Code/QRDUtils.h
Comment (baldurk) [inline qrenderdoc/Code/QRDUtils.h]: This could be just a `class QHeaderView;` declaration instead I think, which makes this header a little lighter to include everywhere.

File: qrenderdoc/Windows/TextureViewer.cpp
Comment (baldurk) [inline qrenderdoc/Windows/TextureViewer.cpp]: Would it be useful here to return the depth for 3D textures, and the array size for non-3D textures? Since they're disjoint (you can't have 3D arrays).

File: qrenderdoc/Windows/TextureViewer.cpp
Comment (baldurk) [inline qrenderdoc/Windows/TextureViewer.cpp]: What's the reason for this TODO, is there a problem with using strings just now? I agree it's a better way than using the builtin `ResourceFormat` sort (which is not really intended to be user-facing).

File: qrenderdoc/Windows/TextureViewer.cpp
Comment (baldurk) [inline qrenderdoc/Windows/TextureViewer.cpp]: Please prefer `NULL` to `nullptr`.

File: qrenderdoc/Windows/TextureViewer.cpp
Comment (baldurk) [inline qrenderdoc/Windows/TextureViewer.cpp]: This isn't well documented but the way the stretch hints are implemented all negative values are treated the same. The logic is that columns with a negative value only get just enough space to fit their contents, and any remaining extra space is assigned to columns with non-negative values, with the actual value indicating the proportion (i.e. if they're all 1 then the extra space is split evenly, if one of them was 2 it would get twice as much extra space as a column with 1).

So this is fine, it determines a layout where only the last column gets extra space which is a common pattern but it'd be clearer to have the other columns all be -1.

ninja edit: Actually I think the first name column should be 1 and the others all -1, because they're logical columns not visual columns. Otherwise if you re-order the columns to put format in the middle somewhere it looks really weird being very wide and crushing the others on the right.

File: qrenderdoc/Windows/TextureViewer.cpp
Comment (baldurk) [inline qrenderdoc/Windows/TextureViewer.cpp]: `Column_TexName` here instead of 0 would be nice.

File: qrenderdoc/Windows/TextureViewer.cpp
Comment (baldurk) [inline qrenderdoc/Windows/TextureViewer.cpp]: Unused variable `oldVisIdx` here.

File: qrenderdoc/Windows/TextureViewer.cpp
Comment (baldurk) [inline qrenderdoc/Windows/TextureViewer.cpp]: Why is this event filter installed here? TextureViewer doesn't implement `eventFilter` so this seems like it wouldn't do anything.

**Review by baldurk — CHANGES_REQUESTED** [review]
This looks like a good change, and you've kind of buried the lede by not explicitly saying that you've added property columns to the texture list which can then be chosen :smile:.

I've left notes about specific bits of the code but a few general points from giving this a spin:

* When I change the sort mode the list doesn't immediately update until I mouse-over the viewport. I think you need to trigger a repaint or something by hand as it doesn't seem to do that by default.
* I think it might be worth trying `std::stable_sort` instead of `std::sort`. I believe that would work better in the UI, so that if you e.g. sort by width descending and then height descending you'd effectively end up with a primary 'sub-sort' of width for equal heights. That tends to be the UI convention for column sorted lists and I think a stable sort is the way it happens (rather than having a deliberately implicit hidden state of previous column sorts, let that fall out of the stable sort).
* The row highlighting looks strange now that there are column dividers. I think we want to follow the pipeline state's model of having an on-hover background colour now instead as that would look better.
* It would be good to expand the filter to search through all columns. There's not much risk I think of unwanted interactions there as the only non-numerical column is the format, and people wanting to search for a texture that contains some format substring won't be surprised if textures with that actual format show up. I can definitely see people wanting to filter/search by only 16-bit textures with a filter of 'R16' or similar.
* Similarly, I'm not sure we want selecting anywhere in the row to immediately open the texture anymore. I can see people wanting to select a row just as habit while browsing, or to select it to copy-paste the details (I think the default RDTreeView copy-paste handler should catch this! though I might be wrong).
  In fact having tried it out a bit in practice I'm increasingly thinking this list should behave the same as the pipeline state lists/tables - with the texture name being a ResourceId QVariant so it becomes a resource-inspector-linked name, and have the go arrow being its own column with a hover hand (it can be excluded from the column selection as an extra column in the model). That would allow selection and then either clicking the arrow or double clicking the row to open the texture.
* Also speculating (this would need to be tested to see how it feels) but since there's a lot more horizontal space used now I wonder if we should rethink the default location of the texture list. For me when I open it on the left side only the texture name column is visible. I wonder if it wouldn't be better to default open it all the way docked on the bottom like so:
  image

**Comment by kevin-mccullough** [comment]
I'm working through your suggestions now.  I'll let you know if I have any questions.

File: qrenderdoc/Code/QRDUtils.cpp
Comment (kevin-mccullough) [reply qrenderdoc/Code/QRDUtils.cpp]: Sorry, I do this internally to track what  we've touched.  I did not mean to imply we were changing the copyright.

**Comment by kevin-mccullough** [comment]
Hi Baldur,
I rebased (-i) this branch with the changes you requested.  Let me know if that is an ok way to do this, or if you prefer that I create a new PR and branch.

File: qrenderdoc/Windows/TextureViewer.cpp
Comment (baldurk) [inline qrenderdoc/Windows/TextureViewer.cpp:2892]: I prefer to have matching {} on if and else - if one needs braces then both should have it. I find it easier to parse matching conditionals at a glance, on top of indentation.

File: qrenderdoc/Windows/TextureViewer.cpp
Comment (baldurk) [inline qrenderdoc/Windows/TextureViewer.cpp]: I do think it would be useful to take the depth to mean either the depth for 3D textures or the array size for non-3D textures. This would also need to be updated to match below in the filtering code.

File: qrenderdoc/Windows/TextureViewer.cpp
Comment (baldurk) [inline qrenderdoc/Windows/TextureViewer.cpp:3952]: I don't think `FilterType::Value` is always valid, though it seems to compile on some of the CI compilers? I'd rather make it explicitly an `enum class` since that also avoids potential collisions with those value names.

File: qrenderdoc/Windows/TextureViewer.cpp
Comment (baldurk) [inline qrenderdoc/Windows/TextureViewer.cpp]: Prefer to avoid `auto` unless the type is really obtuse like stl iterators or lambdas. Here I think just `QVariant` is fine, especially since `QVariant::fromValue` returns a value copy.

File: qrenderdoc/Windows/TextureViewer.h
Comment (baldurk) [inline qrenderdoc/Windows/TextureViewer.h]: Dangling function here? It seems to have no references or implementation.

File: qrenderdoc/Windows/TextureViewer.cpp
Comment (baldurk) [inline qrenderdoc/Windows/TextureViewer.cpp:521]: I don't think we should show the texture list in the default layout here. It's useful but it does take up a good amount of space so I think the list should be default closed and only opened if the user requests it.

File: qrenderdoc/Windows/TextureViewer.cpp
Comment (baldurk) [inline qrenderdoc/Windows/TextureViewer.cpp]: Setting different colours for each row here feels strange since the colours don't correspond to anything. In the pipeline state viewer this is used to show which attributes (possibly several) come from which vertex input bind, since it's a many:one relationship. I think for the texture list it's fine to leave the default RDTreeWidget hover behaviour from calling `setHoverIconColumn`.

File: qrenderdoc/Windows/TextureViewer.cpp
Comment (baldurk) [inline qrenderdoc/Windows/TextureViewer.cpp]: This show/get is not necessary here - we can call directly to `ViewTexture` from inside the texture viewer :smile: .

**Review by baldurk — CHANGES_REQUESTED** [review]
Rebasing and force-pushing is fine, you can also push review feedback as a new commit and then squash/rebase once everything is reviewed and ready to merge. Depends on what feels best for you. Making a new PR for every review feedback wouldn't be good since it would lose track of all the discussion so please don't do that :).

I've left some notes about the refactor, but this is much improved so that's great!

**Comment by baldurk** [comment]
Oh also I forgot - check out the CI failures. I think it's just missing a `class RDTreeWidgetItem` declaration, it looks like it's failing on the platforms that don't have precompiled headers, which can sometimes hide issues like that.

File: qrenderdoc/Windows/TextureViewer.cpp
Comment (kevin-mccullough) [reply qrenderdoc/Windows/TextureViewer.cpp]: I see.  I misunderstood, and previously thought you only wanted the title changed.  How can you tell if a texture is a 3D texture?  If it's depth is 1?

File: qrenderdoc/Windows/TextureViewer.cpp
Comment (kevin-mccullough) [reply qrenderdoc/Windows/TextureViewer.cpp:3952]: I'm not sure I understand.  I will change FilterType to be an `enum class` but it's just defined in TextureViewer.h.  Which names are you worried about collisions with?

File: qrenderdoc/Windows/TextureViewer.h
Comment (kevin-mccullough) [reply qrenderdoc/Windows/TextureViewer.h]: Oh yes sorry, this got renamed.  I'll remove it.

File: qrenderdoc/Windows/TextureViewer.cpp
Comment (kevin-mccullough) [reply qrenderdoc/Windows/TextureViewer.cpp]: 👍

File: qrenderdoc/Windows/TextureViewer.cpp
Comment (baldurk) [reply qrenderdoc/Windows/TextureViewer.cpp]: Yep the title change is good but at the moment this is only filling in the depth, not the slices. Some APIs merge depth and slices but for clarity I keep them separate even if they're mutually exclusive.

`TextureDescription` has a member `dimension` which shows the texture dimension, either 1, 2 or 3.

File: qrenderdoc/Windows/TextureViewer.cpp
Comment (baldurk) [reply qrenderdoc/Windows/TextureViewer.cpp:3952]: Yeh `FilterType` is local to TextureViewer but if it's a plain enum then `String` becomes valid in any class member function. That means it could override a `String` token at global scope - it doesn't leak out but it risks colliding with anything global.

Without it being an `enum class` then the `FilterType::` is optional at best (I actually thought it was illegal but this seems to compile), so generally for this kind of thing I prefer enum class in new code unless casting to/from ints is going to be a real chore.

File: qrenderdoc/Windows/TextureViewer.cpp
Comment (kevin-mccullough) [reply qrenderdoc/Windows/TextureViewer.cpp]: Got it.  I'll make the changes.

File: qrenderdoc/Windows/TextureViewer.cpp
Comment (kevin-mccullough) [reply qrenderdoc/Windows/TextureViewer.cpp:3952]: Got it, thanks!

File: qrenderdoc/Windows/TextureViewer.h
Comment (baldurk) [inline qrenderdoc/Windows/TextureViewer.h]: This is a minor complaint but it's fine to pick a base type for the enum class here, but I'd rather have a more normal one like uint32_t since it's not an issue for storage/perf here :smile: . FYI it's optional to have a base type for enum classes, so you could omit it if you want too.

File: qrenderdoc/Windows/TextureViewer.cpp~RF1f3f6fc6.TMP
Comment (baldurk) [inline qrenderdoc/Windows/TextureViewer.cpp~RF1f3f6fc6.TMP]: This temp file seems to have been added by accident :disappointed: .

**Review by baldurk — APPROVED** [review]

**Pull request merged**
